File: loom-workshop-java/src/test/java/nl/theexperts/loomworkshopjava/service/PipelineServiceTest.java
package nl.theexperts.loomworkshopjava.service;

import nl.theexperts.loomworkshopjava.repository.DataRecord;
import nl.theexperts.loomworkshopjava.repository.DataRecordRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.when;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withStatus;

@ExtendWith(MockitoExtension.class)
class PipelineServiceTest {

    @Mock
    private RestTemplate restTemplate;

    @Mock
    private DataRecordRepository dataRecordRepository;

    @InjectMocks
    private PipelineService pipelineService;

    @Test
    @DisplayName("Test processLargeDataSet performance")
    void testProcessLargeDataSetPerformance() throws ExecutionException, InterruptedException {
        // Mock dependencies
        when(restTemplate.getForObject(any(String.class), eq(String.class))).thenAnswer(request -> {
            try {
                Thread.sleep(300); // Delay
            } catch (InterruptedException ignored) {}
            return withStatus(OK).body("responseBody").contentType(MediaType.APPLICATION_JSON);
        });
        when(dataRecordRepository.findByData(any())).thenReturn(new DataRecord("dbData"));

        List<String> largeInput = new ArrayList<>(Collections.nCopies(300, "test"));

        // Measure time taken
        long startTime = System.currentTimeMillis();
        pipelineService.processLargeDataSet(largeInput);
        long endTime = System.currentTimeMillis();

        long timeTaken = endTime - startTime;
        assertThat(timeTaken).isLessThan(2000L);  // Our slow implementation takes so long! :(
    }
}

File: loom-workshop/src/test/kotlin/nl/theexperts/loomworkshop/service/PipelineServiceTest.kt
package nl.theexperts.loomworkshop.service

import io.mockk.coEvery
import io.mockk.every
import io.mockk.mockk
import nl.theexperts.loomworkshop.repository.DataRecord
import nl.theexperts.loomworkshop.repository.DataRecordRepository
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.web.client.RestTemplate

class PipelineServiceTest {

    private lateinit var restTemplate: RestTemplate
    private lateinit var dataRecordRepository: DataRecordRepository
    private lateinit var pipelineService: PipelineService

    @BeforeEach
    fun setUp() {
        restTemplate = mockk()
        dataRecordRepository = mockk()
        pipelineService = PipelineService(restTemplate, dataRecordRepository)
    }

    @Test
    fun `test processLargeDataSet performance`() {
        // Mock dependencies
        coEvery { restTemplate.getForObject(any<String>(), String::class.java) } coAnswers {
            Thread.sleep(300)  // Simulate a delay
            "externalData"
        }

        every { dataRecordRepository.findByData(any()) } returns DataRecord(data = "dbData")

        val largeInput = List(300) { "input" }

        // Measure time taken
        val startTime = System.currentTimeMillis()
        pipelineService.processLargeDataSet(largeInput)
        val endTime = System.currentTimeMillis()

        val timeTaken = endTime - startTime
        assertTrue(timeTaken < 5000) // Our slow implementation takes so long! :(
    }
}
